src/main/java/org/truffleruby/core/module/ModuleFields.java

```diff
@@ -338,6 +341,13 @@ public void setAutoloadConstant(RubyContext context, Node currentNode, String na
                     autoloadConstant,
                     filename));
         }
+
+        final ReentrantLockFreeingMap<String> fileLocks = getContext().getFeatureLoader().getFileLocks();
+        final ReentrantLock lock = fileLocks.get(filename.toString());
+        if (lock.isLocked()) {
```

Comment:
Should it be `lock.isHeldByCurrentThread()`? Or do you mean to do this even if another thread is loading that path?

What will unlock the lock acquired by `autoloadConstantStart()` which calls `AutoloadConstant#startAutoLoad()`?

This needs a comment explaining the logic here, I don't quite understand it yet.
I guess it's to catch the case of adding an autoload for a file which already started autoloading?
`autoloadConstant` is always a new constant, so this is locking it and the code in RequireNode unlocks it?

I wonder what should happen if there is a file adding an autoload for the same file, but in another Thread. Then unlock would probably fail, right?

Reply:
> Should it be lock.isHeldByCurrentThread()?

Yep, it should be `isHeldByCurrentThread()` :+1:

> What will unlock the lock acquired by autoloadConstantStart() which calls AutoloadConstant#startAutoLoad()?

> This needs a comment explaining the logic here, I don't quite understand it yet.
I guess it's to catch the case of adding an autoload for a file which already started autoloading?
autoloadConstant is always a new constant, so this is locking it and the code in RequireNode unlocks it?

I hope my last comment with explanation of changes answered to these questions.
In short, the lock we add here will be removed in `RequireNode#requireConsideringAutoload` because now we reselect autoload constants and always do it because of `finalize { .. }`.

> I wonder what should happen if there is a file adding an autoload for the same file, but in another Thread. Then unlock would probably fail, right?

I thinks so...at least we have a tagged test for this 
https://github.com/oracle/truffleruby/blob/afa31af4f2879f0dcbc00139cad89bc44bdfbe6b/spec/tags/core/module/autoload_tags.txt#L1

On the other hand, it seems to me that I am not changing the locking behavior here in any way. When this code is called, the file will already be locked at the beginning of `RequireNode#doRequire`. And the blocking of a new constant should not affect anything.

Reply:
Yeah locking seems fine if `lock.isHeldByCurrentThread()` is used (i.e., we know we're the thread autoloading).